pyprocessing/math.py

```diff
@@ -162,191 +176,199 @@ def from_angle(angle, target=None):
+        return sum(a * a for a in self)
 
     def dot(self, other):
-        return self.x * other.x + self.y * other.y + self.z * other.z
+        return sum(a * b for a, b in zip(self, other))
 
     def cross(self, other):
-        x = self.y * other.z - self.z * other.y
-        y = self.z * other.x - self.x * other.z
-        z = self.x * other.y - self.y * other.x
+        if not len(self) == 3:
+            raise ValueError(
+                'Cross product is only defined for 3d vectors'
+            )
+        x = self[1] * other[2] - self[2] * other[1]
+        y = self[2] * other[0] - self[0] * other[2]
+        z = self[0] * other[1] - self[1] * other[0]
         return PVector(x, y, z)
 
     def normalize(self):
-        mag = self.mag()
-        self.x /= mag
-        self.y /= mag
-        self.z /= mag
+        s = self.mag()
+        if s == 0:
+            return
+        for i in range(len(self)):
+            self[i] /= s
+
+    def normalized(self):
+        v = self.copy()
+        v.normalize()
+        return v
 
     def limit(self, limit):
-        if self.mag_sq() <= limit ** 2:
+        s = self.mag_sq()
+        if s <= limit * limit:
             return
 
-        self.normalize()
-        self *= limit
+        s = math.sqrt(s) / limit
+
+        for i in range(len(self)):
+            self[i] /= s
 
     def heading(self):
-        if not self.is_2d:
-            raise ValueError('Cannot calculate heading on a 3D Vector.')
-        return math.atan2(self.y, self.x)
+        return math.atan2(self[1], self[0])
 
     def rotate(self, theta):
-        if not self.is_2d:
-            raise ValueError('Cannot rotate a 3D Vector.')
-        self.x = math.cos(theta) * self.x - math.sin(theta) * self.y
-        self.y = math.sin(theta) * self.x + math.cos(theta) * self.y
+        if len(self) < 2:
+            raise ValueError('Cannot rotate a Vector'
+                             ' with less than 2 elements')
+
+        self[0] = math.cos(theta) * self[0] - math.sin(theta) * self[1]
+        self[1] = math.sin(theta) * self[0] + math.cos(theta) * self[1]
 
     def lerp(self, target, amount):
-        if not (0 <= amount <= 1):
-            raise ValueError('`amount` must be between 0 and 1.')
-        if float(amount) == 0.0:
+        if amount == 0:
             return self
-        if float(amount) == 1.0:
+        if amount == 1:
             return target
-        x = lerp(self.x, target.x, amount)
-        y = lerp(self.y, target.y, amount)
-        z = lerp(self.z, target.z, amount)
-        if self.is_2d and target.is_2d:
-            z = None
-        return PVector(x, y, z)
+        return target * amount + self * (1 - amount)
 
     def angle_between(self, other):
         return math.acos(self.dot(other) / (self.mag() * other.mag()))
 
     def array(self):
-        return self.position
+        return self.position()
+
+    def gen_single_component_access(letter):
```

Comment:
make it private

Reply:
It is being deleted at the end of the execution of the scope, so gen_single_component_access basically dont exist.